File: assignments/week04/basic/PGM_172927_LGJ.java
// PGM 172927 - 광물캐기
// 유형: 그리디/시뮬
// 링크: https://school.programmers.co.kr/learn/courses/30/lessons/172927

import java.util.*;
//picks = [1, 1, 0], minerals = ["iron", "iron", "diamond", "iron", "stone", "diamond", "diamond", "diamond"] 5, 15 / 9, 3

class Solution {
    public int solution(int[] picks, String[] minerals) {
        int answer = 0;
        List<String[]> mineralGroups = new ArrayList<>();

        for (int i=0; i<minerals.length; i += 5) {
            int start = i;
            int end = Math.min(i + 5, minerals.length); // 배열 범위 초과 방지
            String[] mineralGroup = Arrays.copyOfRange(minerals, start, end);
            mineralGroups.add(mineralGroup);
        }
        int totalPickCount = picks[0] + picks[1] + picks[2];
        int groupCount = Math.min(mineralGroups.size(), totalPickCount);
        List<String[]> pickableGroups = mineralGroups.subList(0, groupCount);

        pickableGroups.sort(Comparator.comparingInt(Solution::sumOfMineralGroup).reversed());


        for (String[] mineralGroup : pickableGroups) {
            System.out.println(Arrays.toString(mineralGroup));
            if (picks[0] > 0) {
                for (String mineral : mineralGroup) {
                    answer += getMiningCost(mineral, "diamond");
                }
                picks[0] -= 1;
            } else if (picks[1] > 0) {
                for (String mineral : mineralGroup) {
                    answer += getMiningCost(mineral, "iron");
                }
                picks[1] -= 1;
            } else if (picks[2] > 0) {
                for (String mineral : mineralGroup) {
                    answer += getMiningCost(mineral, "stone");
                }
                picks[2] -= 1;
            } else {
                break;
            }
        }

        return answer;
    }
    // 가장 캐기 힘든 그룹(=stone 곡괭이 기준 피로도 높은 순)부터 정렬
    private static int sumOfMineralGroup(String[] mineralGroup) {
        int sum = 0;
        HashMap<String, Integer> map = new HashMap<>();
        map.put("diamond", 25);
        map.put("iron", 5);
        map.put("stone", 1);

        for (String mineral : mineralGroup) {
            sum += map.get(mineral);
        }
        return sum;
    }


    // 광물 + 곡괭이 조합에 따른 캐기 비용
    private static int getMiningCost(String mineral, String pick) {
        if (pick.equals("diamond")) {
            return 1;
        } else if (pick.equals("iron")) {
            if (mineral.equals("diamond")) return 5;
            else return 1;
        } else if (pick.equals("stone")) {
            if (mineral.equals("diamond")) return 25;
            else if (mineral.equals("iron")) return 5;
            else return 1;
        }
        return 0;
    }

}

File: assignments/week04/basic/PGM_172927_LGJ.py
# PGM 172927 - 광물캐기
# 유형: 그리디/시뮬
# 링크: https://school.programmers.co.kr/learn/courses/30/lessons/172927


def solution(picks, minerals):
    answer = 0
    minerals_dict = {'diamond': {0:1, 1:5, 2:25}, 'iron': {0:1, 1:1, 2:5}, 'stone': {0:1, 1:1, 2:1}}
    # 비싼 원석들이 많은 그룹 순으로 정렬
    sorted_groups = sort_minerals_by_fatigue_stone(picks, minerals)
    
    for group in sorted_groups: # 최소 피로도 > 비싼 원석 그룹은 좋은 장비 순으로 
        if picks[0]: # diamond 곡괭이
            for mineral in group:
                answer += minerals_dict[mineral][0]
            picks[0] -= 1
        elif picks[1]: # iron 곡괭이
            for mineral in group:
                answer += minerals_dict[mineral][1]
            picks[1] -= 1
        elif picks[2]: # stone 곡괭이
            for mineral in group:
                 answer += minerals_dict[mineral][2]
            picks[2] -= 1
        else: # 곡괭이 소진
            break

    return answer

def fatigue_stone(mineral_group): # 광물 그룹 별 피로도의 합
    fatigue_table = {'diamond': 25, 'iron': 5, 'stone': 1} # 각 광물의 최대 피로도
    return sum(fatigue_table[mineral] for mineral in mineral_group)

def sort_minerals_by_fatigue_stone(picks, minerals): # 5개씩 그룹화 및 피로도가 높은 순 정렬
    max_minerals = sum(picks) * 5
    minerals = minerals[:max_minerals]  
    grouped_minerals = [minerals[i:i+5] for i in range(0, len(minerals), 5)]#[dia, iron, stone], [stone, stone, stone]]
    
    sorted_groups = sorted(grouped_minerals, key=fatigue_stone, reverse=True) 
    return sorted_groups
